File: app/src/main/java/com/sg/moviesindex/service/network/TMDbService.java
package com.sg.moviesindex.service.network;

import com.sg.moviesindex.model.tmdb.CastsList;
import com.sg.moviesindex.model.tmdb.DiscoversList;
import com.sg.moviesindex.model.tmdb.GenresList;
import com.sg.moviesindex.model.tmdb.Movie;
import com.sg.moviesindex.model.tmdb.MoviesList;
import com.sg.moviesindex.model.tmdb.ReviewsList;

import io.reactivex.Observable;
import retrofit2.http.GET;
import retrofit2.http.Path;
import retrofit2.http.Query;

public interface TMDbService {

    //@GET("movie/popular")
    //Call<MovieDBResponse> getPopularMovies(@Query("api_key")String apiKey, @Query("page")int pageIndex);

    @GET("movie/popular")
    Observable<MoviesList> getPopularMoviesWithRx(@Query("api_key") String apiKey, @Query("page") int pageIndex);

    @GET("movie/top_rated")
    Observable<MoviesList> getTopRatedMoviesWithRx(@Query("api_key") String apiKey, @Query("page") int pageIndex);

    @GET("movie/upcoming")
    Observable<MoviesList> getUpcomingMoviesWithRx(@Query("api_key") String apiKey, @Query("page") int pageIndex, @Query("region") String region);

    @GET("movie/now_playing")
    Observable<MoviesList> getNowPlayingWithRx(@Query("api_key") String apiKey, @Query("page") int pageIndex, @Query("region") String region);

    @GET("genre/movie/list")
    Observable<GenresList> getGenresList(@Query("api_key") String apiKey);

    @GET("movie/{movieId}/reviews")
    Observable<ReviewsList> getReviews(@Path("movieId") Long movieId, @Query("api_key") String apiKey, @Query("page") int pageIndex);

    @GET("movie/{movieId}/credits")
    Observable<CastsList> getCasts(@Path("movieId") Long movieId, @Query("api_key") String apiKey);

    @GET("discover/movie")
    Observable<DiscoversList> discover(@Query("api_key") String apiKey, @Query("with_genres") String genres, @Query("include_adult") Boolean adult,
                                       @Query("include_video") Boolean video, @Query("page") int pageIndex, @Query("sort_by") String sortBy,
                                       @Query("region") String region, @Query("with_original_language") String language,
                                       @Query("release_date.gte") String releaseDateGTE, @Query("release_date.lte") String releaseDateLTE);

    @GET("search/movie")
    Observable<DiscoversList> search(@Query("api_key") String apiKey, @Query("include_adult") Boolean adult, @Query("query") String query, @Query("page") int page);

    @GET("movie/{id}")
    Observable<Movie> getFullMovieInformation(@Path("id") Long id, @Query("api_key") String apiKey);


}

File: app/src/main/java/com/sg/moviesindex/data/remote/TMDbService.kt
package com.sg.moviesindex.data.remote

import com.sg.moviesindex.data.local.Movie
import io.reactivex.Observable
import retrofit2.http.GET
import retrofit2.http.Path
import retrofit2.http.Query

/**
 * Retrofit interface for the TMDb (The Movie Database) API.
 * Defines the endpoints and their required parameters for fetching movie-related data.
 */
interface TMDbService {
  @GET("movie/popular")
  fun getPopularMoviesWithRx(
    @Query("api_key") apiKey: String,
    @Query("page") pageIndex: Int,
  ): Observable<MoviesList>

  @GET("movie/top_rated")
  fun getTopRatedMoviesWithRx(
    @Query("api_key") apiKey: String,
    @Query("page") pageIndex: Int,
  ): Observable<MoviesList>

  @GET("movie/upcoming")
  fun getUpcomingMoviesWithRx(
    @Query("api_key") apiKey: String,
    @Query("page") pageIndex: Int,
    @Query("region") region: String?,
  ): Observable<MoviesList>

  @GET("movie/now_playing")
  fun getNowPlayingWithRx(
    @Query("api_key") apiKey: String,
    @Query("page") pageIndex: Int,
    @Query("region") region: String?,
  ): Observable<MoviesList>

  @GET("genre/movie/list")
  fun getGenresList(
    @Query("api_key") apiKey: String,
  ): Observable<GenresList>

  @GET("movie/{movieId}/reviews")
  fun getReviews(
    @Path("movieId") movieId: Long,
    @Query("api_key") apiKey: String,
    @Query("page") pageIndex: Int,
  ): Observable<ReviewsList>

  @GET("movie/{movieId}/credits")
  fun getCasts(
    @Path("movieId") movieId: Long,
    @Query("api_key") apiKey: String,
  ): Observable<CastsList>

  @GET("discover/movie")
  fun discover(
    @Query("api_key") apiKey: String,
    @Query("with_genres") genres: String?,
    @Query("include_adult") adult: Boolean?,
    @Query("include_video") video: Boolean?,
    @Query("page") pageIndex: Int,
    @Query("sort_by") sortBy: String?,
    @Query("region") region: String?,
    @Query("with_original_language") language: String?,
    @Query("release_date.gte") releaseDateGTE: String?,
    @Query("release_date.lte") releaseDateLTE: String?,
  ): Observable<DiscoversList>

  @GET("search/movie")
  fun search(
    @Query("api_key") apiKey: String,
    @Query("include_adult") adult: Boolean?,
    @Query("query") query: String,
    @Query("page") page: Int,
  ): Observable<DiscoversList>

  @GET("movie/{id}")
  fun getFullMovieInformation(
    @Path("id") id: Long,
    @Query("api_key") apiKey: String,
  ): Observable<Movie>
}
